### PR TITLE
[Back] Add friends functionality to bot

### PR DESCRIPTION
**Pull request by animepunkgirl**
1. Added ignoring .env in sub directories to .gitignore
2. Transfered commands and callbacks factories to separate modules
3. Added callbacks (telegram' callback query handlers) to bot bootstrap
4. Changed command's handle / isMatching message parameter to MetaMessage (message + message's meta)
5. Added addCallbackQueryListener, getChat, getFileLink, answerCallbackQuery, deleteMessage methods to BotService
6. Added DeleteCallback (handles the request to remove a friend)
7. Added RequestCallback (handles accepting/declining incoming friend request)
8. Added AddCommand (handles adding another user to friend list)
9. Added FriendsCommand (shows user's friend list with Delete button)
10. Removed unnecessary isMatching from IncorrectCommand
11. Added RequestsCommand (shows user's incoming friend requests)
12. Now when calling the /start command, a user is created in the database
13. Added Notifier Module and Service (contains methods that are called when a user needs to be notified of another user's actions (you have been friended / you have been deleted from friends)
14. Added Friend Request Module and Service (contains methods to CRUD Friend Requests)
15. Fixed bug with Logger debug method
16. Added AppLogger that logs Nest logs
17. Added Friend Request Model to Database
18. Added required and unique keywords to User Model' chat_id, added username and friends fields to User Model
19. Now when user authorize through webapp we update their username if it's changed
20. Added getFriends route to user controller
21. Added getUserByUsername, getFriendsByIds, deleteFriend, updateUsername, isUserAreFriends methods to UserService
22. Added typing

### CODE REVIEW

File: bot/src/bot/callback/request.callback.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  constructor(private friendRequestService: FriendRequestService, private botService: BotService) {
+    super();
+  }
+
+  // TODO: Notify friend request sender
+  async handle(callbackQuery: TelegramBot.CallbackQuery): Promise<void> {
+    if(!callbackQuery.data)
+      return;
+
+    const data = callbackQuery.data.substring(this.name.length+1).split('_');
```
Comment (omgdendi) [inline bot/src/bot/callback/request.callback.ts:20]: put it in a separate function

File: bot/src/bot/command/requests.command.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import {Command} from "./command";
+import {MetaMessage} from "../bot.types";
+import {UserService} from "../../user/user.service";
+import {FriendRequestService} from "../../friend-request/friend-request.service";
+import TelegramBot from "node-telegram-bot-api";
+import {BotService} from "../bot.service";
+import {FriendRequest, FriendRequestDocument} from "../../schemas/friend-request.schema";
+import {Types} from "mongoose";
+
+// TODO: Show list of incoming friend requests
```
Comment (omgdendi) [inline bot/src/bot/command/requests.command.ts:11]: remove TODO

File: bot/src/bot/command/requests.command.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  private async showRequests(chatId: TelegramBot.ChatId, requests: FriendRequestDocument[]) {
+    for await (const request of requests) {
+      await this.showRequest(chatId, request)
+    }
+  }
+
+  private async showRequest(chatId: TelegramBot.ChatId, request: FriendRequestDocument) {
+    const fromId = new Types.ObjectId(request.from._id)
+    const from = await this.userService.getUserById(new Types.ObjectId(fromId))
```
Comment (omgdendi) [inline bot/src/bot/command/requests.command.ts:41]: remove unnecessary second new ObjectId

File: bot/src/friend-request/friend-request.service.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  // TODO: Catch errors
+  async create(from: UserDocument["_id"], to: UserDocument["_id"]): Promise<FriendRequestDocument> {
+    const existedFriendRequest = await this.friendRequestModel.findOne({
+      from, to
+    }).exec()
+    if(existedFriendRequest)
+      throw new Error('Friend request already exists')
+
+    const isUserAreFriends = await this.userService.isUserAreFriends(from, to);
```
Comment (omgdendi) [inline bot/src/friend-request/friend-request.service.ts:23]: rename to isUsersAreFriends

File: bot/src/friend-request/friend-request.service.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  async accept(id: string) {
+    const friendRequest = await this.friendRequestModel
+      .findById(id)
+      .populate('from')
+      .populate('to')
+      .exec()
+    if(!friendRequest)
+      return;
+
+    const from = friendRequest.from;
```
Comment (omgdendi) [inline bot/src/friend-request/friend-request.service.ts:43]: put add friends to separate function

File: bot/src/user/user.controller.ts [hunk 6]
```diff
@@ -17,13 +23,35 @@ export class UserController {
...
   }
 
+  @Auth()
+  @Get('/friends')
+  async getFriends(@Me() user: User): Promise<FriendsDto> {
+    const friends = await this.userService.getFriendsByIds(user.friends)
+    const friendList: Friend[] = []
+    for await (const friend of friends) {
+      const info = await this.botService.getChat(friend.chat_id)
+      const avatar_url = info.photo ? await this.botService.getFileLink(info.photo.small_file_id) : undefined
```
Comment (omgdendi) [inline bot/src/user/user.controller.ts:43]: transfer checking fileId to getFileLink(), this should return null if it doesn't have fileId

File: bot/src/user/user.service.ts [hunk 7]
```diff
@@ -54,6 +60,75 @@ export class UserService {
...
+
+    const info = await this.botService.getChat(chat_id)
+    if(!info.username)
+      return;
+
+    user.username = info.username
+    await user.save()
+  }
+
+  async isUserAreFriends(first: Types.ObjectId, second: Types.ObjectId): Promise<boolean> {
```
Comment (omgdendi) [inline bot/src/user/user.service.ts:122]: rename to isUsersAreFriends

File: bot/src/bot/callback/request.callback.ts (hunk 1 above)
Comment (animepunkgirl) [reply bot/src/bot/callback/request.callback.ts:20]: Let's do it when we'll refactor this methods

**Pull request merged**